core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs

```diff
@@ -0,0 +1,249 @@
+    ) -> anyhow::Result<Option<PgOrRocksdbStorage<'a>>>;
+}
+
+/// A [`ReadStorage`] implementation that uses either [`PostgresStorage`] or [`RocksdbStorage`]
+/// underneath.
+#[derive(Debug)]
+pub enum PgOrRocksdbStorage<'a> {
+    Postgres(PostgresStorage<'a>),
+    Rocksdb(RocksdbStorage),
+}
+
+impl ReadStorage for PgOrRocksdbStorage<'_> {
+    fn read_value(&mut self, key: &zksync_types::StorageKey) -> zksync_types::StorageValue {
+        match self {
+            Self::Postgres(postgres) => postgres.read_value(key),
+            Self::Rocksdb(rocksdb) => rocksdb.read_value(key),
+        }
+    }
+
+    fn is_write_initial(&mut self, key: &zksync_types::StorageKey) -> bool {
+        match self {
+            Self::Postgres(postgres) => postgres.is_write_initial(key),
+            Self::Rocksdb(rocksdb) => rocksdb.is_write_initial(key),
+        }
+    }
+
+    fn load_factory_dep(&mut self, hash: zksync_types::H256) -> Option<Vec<u8>> {
+        match self {
+            Self::Postgres(postgres) => postgres.load_factory_dep(hash),
+            Self::Rocksdb(rocksdb) => rocksdb.load_factory_dep(hash),
+        }
+    }
+
+    fn get_enumeration_index(&mut self, key: &zksync_types::StorageKey) -> Option<u64> {
+        match self {
+            Self::Postgres(postgres) => postgres.get_enumeration_index(key),
+            Self::Rocksdb(rocksdb) => rocksdb.get_enumeration_index(key),
+        }
+    }
+}
+
+impl<'a> From<PostgresStorage<'a>> for PgOrRocksdbStorage<'a> {
+    fn from(value: PostgresStorage<'a>) -> Self {
+        Self::Postgres(value)
+    }
+}
+
+impl<'a> From<RocksdbStorage> for PgOrRocksdbStorage<'a> {
+    fn from(value: RocksdbStorage) -> Self {
+        Self::Rocksdb(value)
+    }
+}
+
+/// A [`ReadStorageFactory`] implementation that can produce short-lived [`ReadStorage`] handles
+/// backed by either Postgres or RocksDB (if it's caught up). Always initialized as a `Postgres`
+/// variant and is then mutated into `Rocksdb` once RocksDB cache is caught up. After which it
+/// can never revert back to `Postgres` as we assume RocksDB cannot fall behind under normal state
+/// keeper operation.
+#[derive(Debug, Clone)]
+pub struct AsyncRocksdbCache {
+    pool: ConnectionPool,
+    rocksdb_cell: Arc<OnceCell<RocksDB<StateKeeperColumnFamily>>>,
+}
+
+impl AsyncRocksdbCache {
+    /// Load latest sealed miniblock from the latest sealed L1 batch (ignores sealed miniblocks
+    /// from unsealed batches).
+    async fn load_latest_sealed_miniblock(
+        connection: &mut StorageProcessor<'_>,
+    ) -> anyhow::Result<Option<(MiniblockNumber, L1BatchNumber)>> {
+        let mut dal = connection.blocks_dal();
+        let Some(l1_batch_number) = dal.get_sealed_l1_batch_number().await? else {
+            return Ok(None);
+        };
+        let Some((_, miniblock_number)) =
+            dal.get_miniblock_range_of_l1_batch(l1_batch_number).await?
+        else {
+            return Ok(None);
+        };
```

Comment:
Nit: I'd consider returning `get_miniblock_range_of_l1_batch` returning `Ok(None)` an error here because the L1 batch is guaranteed to exist. So, it would make sense to convert it to an error using `.context(_)`.